runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/base/tree.h"
+
+#include "iree/base/allocator.h"
+#include "iree/base/assert.h"
+#include "iree/base/config.h"
+#include "stdint.h"
+
+static void iree_tree_add_node_to_cache(iree_tree_t* tree,
+                                        iree_tree_node_t* node);
+
+iree_status_t iree_tree_initialize(iree_allocator_t allocator,
+                                   iree_host_size_t element_size,
+                                   void* initial_node_cache,
+                                   iree_host_size_t initial_node_cache_size,
+                                   iree_tree_t* tree) {
+  tree->element_size = element_size;
+  tree->allocator = allocator;
+  tree->root = &tree->nil;
+  tree->size = 0;
+  tree->cache = NULL;  // Initialize cache
+  memset(&tree->nil, 0x00, sizeof(iree_tree_node_t));
+  tree->nil.is_sentinel = true;
+  tree->initial_node_cache = initial_node_cache;
+  tree->initial_node_cache_size = initial_node_cache_size;
+  if (initial_node_cache) {
+    memset(initial_node_cache, 0, initial_node_cache_size);
+    iree_host_size_t node_size =
+        iree_host_align(sizeof(iree_tree_node_t) + element_size, 16);
+
+    iree_tree_node_t* node = (iree_tree_node_t*)initial_node_cache;
+    for (iree_host_size_t i = 0; i < initial_node_cache_size / node_size; i++) {
+      node->data = (uint8_t*)node + sizeof(iree_tree_node_t);
+      iree_tree_add_node_to_cache(tree, node);
+      node = (iree_tree_node_t*)((uint8_t*)node + node_size);
+    }
+  }
+  return iree_ok_status();
```

Comment:
if a function can never fail you can remove the status return (unless it's a public API and it may be able to fail in the future, in which case having it is better as callers won't have to worry about it magically getting a status - easier to get compile errors when going return -> void than void -> return)